File: brdgme-go/roll_through_the_ages/command.go
package roll_through_the_ages

import (
	"github.com/brdgme/brdgme/brdgme-go/brdgme"
)

type BuildType = int

const (
	BuildTypeCity BuildType = iota
	BuildTypeShip
	BuildTypeMonument
)

type BuildCommand struct {
	Amount int
	Target BuildCommandTarget
}

type BuildCommandTarget struct {
	Type     BuildType
	Monument MonumentID
}

type BuyCommand struct {
	Development DevelopmentID
	Goods       BuyCommandGoods
}

type BuyCommandGoods struct {
	AllGoods bool
	Goods    []Good
}

type TradeCommand struct {
	Amount int
}

type NextCommand struct{}

type TakeCommand struct {
	Actions []TakeAction
}

type DiscardCommand struct {
	Amount int
	Good   Good
}

type InvadeCommand struct {
	Amount int
}

type RollCommand struct {
	Dice []int
}

type SellCommand struct {
	Amount int
}

func (g *Game) CommandParser(player int) brdgme.Parser {
	parsers := []brdgme.Parser{}
	if g.CanBuild(player) {
		parsers = append(parsers, g.BuildParser(player))
	}
	if g.CanBuy(player) {
		parsers = append(parsers, g.BuyParser(player))
	}
	if g.CanTrade(player) {
		parsers = append(parsers, g.TradeParser(player))
	}
	if g.CanNext(player) {
		parsers = append(parsers, g.NextParser())
	}
	if g.CanTake(player) {
		parsers = append(parsers, g.TakeParser(player))
	}
	if g.CanDiscard(player) {
		parsers = append(parsers, g.DiscardParser(player))
	}
	if g.CanInvade(player) {
		parsers = append(parsers, g.InvadeParser(player))
	}
	if g.CanRoll(player) {
		parsers = append(parsers, g.RollParser(player))
	}
	if g.CanSell(player) {
		parsers = append(parsers, g.SellParser(player))
	}
	return brdgme.OneOf(parsers)
}

func (g *Game) BuildParser(player int) brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "build",
				Desc:   "build a city, monument or ship",
				Parser: brdgme.Token("build"),
			},
			brdgme.AfterSpace(
				brdgme.OneOf([]brdgme.Parser{
					g.BuildTargetWorkerParser(player),
					g.BuildTargetShipParser(player),
				}),
			),
		},
		Func: func(value interface{}) interface{} {
			return value.([]interface{})[1].(BuildCommand)
		},
	}
}

func (g *Game) BuildTargetWorkerParser(player int) brdgme.Parser {
	min := 1
	max := g.RemainingWorkers

	parsers := []brdgme.Parser{}
	if g.Boards[player].CityProgress < MaxCityProgress {
		parsers = append(parsers, BuildTargetCityParser)
	}
	if len(g.AvailableMonuments(player)) > 0 {
		parsers = append(parsers, g.BuildTargetMonumentParser(player))
	}

	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name: "amount",
				Desc: "the amount to build",
				Parser: brdgme.Int{
					Min: &min,
					Max: &max,
				},
			},
			brdgme.AfterSpace(brdgme.OneOf(parsers)),
		},
		Func: func(value interface{}) interface{} {
			return BuildCommand{
				Amount: value.([]interface{})[0].(int),
				Target: value.([]interface{})[1].(BuildCommandTarget),
			}
		},
	}
}

var BuildTargetCityParser = brdgme.Map{
	Parser: brdgme.Token("city"),
	Func: func(value interface{}) interface{} {
		return BuildCommandTarget{
			Type: BuildTypeCity,
		}
	},
}

func (g *Game) BuildTargetShipParser(player int) brdgme.Parser {
	min := 1
	max := g.Boards[player].Goods[GoodWood]
	if max < g.Boards[player].Goods[GoodCloth] {
		max = g.Boards[player].Goods[GoodCloth]
	}

	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name: "amount",
				Desc: "the amount of ships to build",
				Parser: brdgme.Int{
					Min: &min,
					Max: &max,
				},
			},
			brdgme.AfterSpace(brdgme.Token("ship")),
		},
		Func: func(value interface{}) interface{} {
			return BuildCommand{
				Amount: value.([]interface{})[0].(int),
				Target: BuildCommandTarget{
					Type: BuildTypeShip,
				},
			}
		},
	}
}

func (g *Game) BuildTargetMonumentParser(player int) brdgme.Parser {
	values := []brdgme.EnumValue{}
	for _, m := range g.AvailableMonuments(player) {
		values = append(values, brdgme.EnumValue{
			Name:  MonumentValues[m].Name,
			Value: m,
		})
	}
	return brdgme.Map{
		Parser: brdgme.Doc{
			Name: "monument",
			Desc: "the monument to build",
			Parser: brdgme.Enum{
				Values: values,
			},
		},
		Func: func(value interface{}) interface{} {
			return BuildCommandTarget{
				Type:     BuildTypeMonument,
				Monument: value.(MonumentID),
			}
		},
	}
}

func (g *Game) TradeParser(player int) brdgme.Parser {
	min := 1
	max := g.Boards[player].Goods[GoodStone]
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "trade",
				Desc:   "trade stone for 3 workers each",
				Parser: brdgme.Token("trade"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "amount",
					Desc: "the amount of stone to trade",
					Parser: brdgme.Int{
						Min: &min,
						Max: &max,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return TradeCommand{
				Amount: value.([]interface{})[1].(int),
			}
		},
	}
}

func (g *Game) BuyParser(player int) brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "buy",
				Desc:   "buy a development",
				Parser: brdgme.Token("buy"),
			},
			brdgme.AfterSpace(
				g.BuyDevelopmentParser(player),
			),
			brdgme.Opt{
				Parser: brdgme.AfterSpace(g.BuyGoodParser(player)),
			},
		},
		Func: func(value interface{}) interface{} {
			goods := BuyCommandGoods{}
			parsedGoods := value.([]interface{})[2]
			if parsedGoods != nil {
				goods = parsedGoods.(BuyCommandGoods)
			}
			return BuyCommand{
				Development: value.([]interface{})[1].(DevelopmentID),
				Goods:       goods,
			}
		},
	}
}

func (g *Game) BuyDevelopmentParser(player int) brdgme.Parser {
	values := []brdgme.EnumValue{}
	for _, d := range g.AvailableDevelopments(player) {
		values = append(values, brdgme.EnumValue{
			Name:  DevelopmentValues[d].Name,
			Value: d,
		})
	}
	return brdgme.Map{
		Parser: brdgme.Doc{
			Name: "development",
			Desc: "the development to buy",
			Parser: brdgme.Enum{
				Values: values,
			},
		},
		Func: func(value interface{}) interface{} {
			return value.(DevelopmentID)
		},
	}
}

func (g *Game) BuyGoodParser(player int) brdgme.Parser {
	var min uint = 1
	return brdgme.OneOf{
		brdgme.Map{
			Parser: brdgme.Token("all"),
			Func: func(value interface{}) interface{} {
				return BuyCommandGoods{
					AllGoods: true,
				}
			},
		},
		brdgme.Map{
			Parser: brdgme.Many{
				Min:    &min,
				Parser: GoodParser(),
			},
			Func: func(value interface{}) interface{} {
				goods := []Good{}
				for _, g := range value.([]interface{}) {
					goods = append(goods, g.(Good))
				}
				return BuyCommandGoods{
					Goods: goods,
				}
			},
		},
	}
}

func GoodParser() brdgme.Parser {
	values := []brdgme.EnumValue{}
	for _, good := range Goods {
		values = append(values, brdgme.EnumValue{
			Name:  GoodStrings[good],
			Value: good,
		})
	}
	return brdgme.Enum{
		Values: values,
	}
}

func (g *Game) NextParser() brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Doc{
			Name:   "next",
			Desc:   "continue to the next phase of your turn",
			Parser: brdgme.Token("next"),
		},
		Func: func(value interface{}) interface{} {
			return NextCommand{}
		},
	}
}

func (g *Game) TakeParser(player int) brdgme.Parser {
	var min uint = 1
	var max uint = 0
	for _, v := range g.KeptDice {
		if v == DiceFoodOrWorkers {
			max += 1
		}
	}
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "take",
				Desc:   "take food or workers from dice",
				Parser: brdgme.Token("take"),
			},
			brdgme.AfterSpace(
				brdgme.Many{
					Min: &min,
					Max: &max,
					Parser: brdgme.Enum{
						Values: []brdgme.EnumValue{
							{Name: TakeMap[TakeFood], Value: TakeFood},
							{Name: TakeMap[TakeWorkers], Value: TakeWorkers},
						},
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			actions := []TakeAction{}
			for _, v := range value.([]interface{})[1].([]interface{}) {
				actions = append(actions, v.(TakeAction))
			}
			return TakeCommand{Actions: actions}
		},
	}
}

func (g *Game) DiscardParser(player int) brdgme.Parser {
	min := 1
	max := g.Boards[player].GoodsOverLimit()
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "discard",
				Desc:   "discard goods down to the limit",
				Parser: brdgme.Token("discard"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "amount",
					Desc: "amount of goods to discard",
					Parser: brdgme.Int{
						Min: &min,
						Max: &max,
					},
				},
			),
			brdgme.AfterSpace(
				brdgme.Doc{
					Name:   "good",
					Desc:   "type of good to discard",
					Parser: GoodParser(),
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return DiscardCommand{
				Amount: value.([]interface{})[1].(int),
				Good:   value.([]interface{})[2].(Good),
			}
		},
	}
}

func (g *Game) InvadeParser(player int) brdgme.Parser {
	min := 1
	max := g.Boards[player].Goods[GoodSpearhead]
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "invade",
				Desc:   "invade other players by spending spearheads for -2 points each",
				Parser: brdgme.Token("invade"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "amount",
					Desc: "amount of spearheads to spend",
					Parser: brdgme.Int{
						Min: &min,
						Max: &max,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return InvadeCommand{
				Amount: value.([]interface{})[1].(int),
			}
		},
	}
}

func (g *Game) RollParser(player int) brdgme.Parser {
	minI := 1
	maxI := len(g.RolledDice)
	minU := uint(minI)
	maxU := uint(maxI)
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "roll",
				Desc:   "roll dice",
				Parser: brdgme.Token("roll"),
			},
			brdgme.AfterSpace(
				brdgme.Many{
					Min: &minU,
					Max: &maxU,
					Parser: brdgme.Int{
						Min: &minI,
						Max: &maxI,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			dice := []int{}
			for _, v := range value.([]interface{})[1].([]interface{}) {
				dice = append(dice, v.(int))
			}
			return RollCommand{Dice: dice}
		},
	}
}

func (g *Game) SellParser(player int) brdgme.Parser {
	min := 1
	max := g.Boards[player].Food
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "sell",
				Desc:   "sell food for 6 coins each",
				Parser: brdgme.Token("sell"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "amount",
					Desc: "amount of food to sell",
					Parser: brdgme.Int{
						Min: &min,
						Max: &max,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return SellCommand{
				Amount: value.([]interface{})[1].(int),
			}
		},
	}
}

File: rust/game/roll-through-the-ages-2/src/command.rs
//! Port of `brdgme-go/roll_through_the_ages_1/command.go`.
//!
//! All 11 gated sub-parsers were wired up in Task 2, matching
//! `CommandParser`'s `OneOf` construction. Task 3 wires the remaining 8
//! `Command` variants (`Build`/`Trade`/`Buy`/`Take`/`Discard`/`Invade`/
//! `Sell`/`Swap`) to real actions in `lib.rs`'s `command()` dispatch.

use brdgme_game::command::parser::*;

use crate::development::DevelopmentId;
use crate::good::{GOODS, Good};
use crate::monument::MonumentId;
use crate::take::TakeAction;
use crate::{Game, player_board};

#[derive(Debug, PartialEq, Clone)]
pub enum BuildTarget {
    City,
    Ship,
    Monument(MonumentId),
}

#[derive(Debug, PartialEq, Clone, Default)]
pub struct BuyGoods {
    pub all_goods: bool,
    pub goods: Vec<Good>,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Command {
    Next,
    Roll {
        dice: Vec<i32>,
    },
    Preserve,
    Build {
        amount: i32,
        target: BuildTarget,
    },
    Trade {
        amount: i32,
    },
    Buy {
        development: DevelopmentId,
        goods: BuyGoods,
    },
    Take {
        actions: Vec<TakeAction>,
    },
    Discard {
        amount: i32,
        good: Good,
    },
    Invade {
        amount: i32,
    },
    Sell {
        amount: i32,
    },
    Swap {
        amount: i32,
        from: Good,
        to: Good,
    },
}

/// Wrapper so `DevelopmentId`/`MonumentId`/`Good`/`TakeAction` (which don't
/// implement `Display` themselves, to keep the domain modules free of
/// parser-layer concerns) can be used with the `Enum` parser, which
/// requires `ToString + Clone`. Port of the ad-hoc `brdgme.EnumValue`
/// name/value pairs built in each Go `*Parser` function.
#[derive(Debug, Clone, Copy)]
struct Choice<T> {
    name: &'static str,
    value: T,
}

impl<T> std::fmt::Display for Choice<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

fn good_choices(goods: &[Good]) -> Vec<Choice<Good>> {
    goods
        .iter()
        .map(|&g| Choice {
            name: g.name(),
            value: g,
        })
        .collect()
}

/// Port of `GoodParser`.
fn good_parser() -> impl Parser<T = Good> {
    Map::new(Enum::partial(good_choices(&GOODS)), |c: Choice<Good>| {
        c.value
    })
}

impl Game {
    pub fn command_parser(&self, player: usize) -> Option<Box<dyn Parser<T = Command> + '_>> {
        let mut parsers: Vec<Box<dyn Parser<T = Command>>> = vec![];
        if self.can_build(player) {
            parsers.push(Box::new(self.build_parser(player)));
        }
        if self.can_buy(player) {
            parsers.push(Box::new(self.buy_parser(player)));
        }
        if self.can_trade(player) {
            parsers.push(Box::new(self.trade_parser(player)));
        }
        if self.can_next(player) {
            parsers.push(Box::new(next_parser()));
        }
        if self.can_take(player) {
            parsers.push(Box::new(self.take_parser()));
        }
        if self.can_discard(player) {
            parsers.push(Box::new(self.discard_parser(player)));
        }
        if self.can_invade(player) {
            parsers.push(Box::new(self.invade_parser(player)));
        }
        if self.can_roll(player) {
            parsers.push(Box::new(self.roll_parser()));
        }
        if self.can_sell(player) {
            parsers.push(Box::new(self.sell_parser(player)));
        }
        if self.can_preserve(player) {
            parsers.push(Box::new(preserve_parser()));
        }
        if self.can_swap(player) {
            parsers.push(Box::new(swap_parser()));
        }
        if parsers.is_empty() {
            None
        } else {
            Some(Box::new(OneOf::new(parsers)))
        }
    }

    /// Port of `BuildParser`/`BuildTargetWorkerParser`/
    /// `BuildTargetShipParser`/`BuildTargetMonumentParser`.
    fn build_parser(&self, player: usize) -> impl Parser<T = Command> {
        let b = &self.boards[player];

        // BuildTargetWorkerParser: `<amount> city|<monument>`.
        let worker_variant: Option<Box<dyn Parser<T = Command>>> =
            if self.can_build_building(player) {
                let mut target_opts: Vec<Box<dyn Parser<T = BuildTarget>>> = vec![];
                if b.city_progress < player_board::MAX_CITY_PROGRESS {
                    target_opts.push(Box::new(Map::new(Token::new("city"), |_| {
                        BuildTarget::City
                    })));
                }
                let available_monuments = self.available_monuments(player);
                if !available_monuments.is_empty() {
                    let choices: Vec<Choice<MonumentId>> = available_monuments
                        .iter()
                        .map(|&m| Choice {
                            name: m.value().name,
                            value: m,
                        })
                        .collect();
                    target_opts.push(Box::new(Map::new(
                        Doc::name_desc("monument", "the monument to build", Enum::partial(choices)),
                        |c: Choice<MonumentId>| BuildTarget::Monument(c.value),
                    )));
                }
                let max = self.remaining_workers;
                Some(Box::new(Map::new(
                    Chain2::new(
                        Doc::name_desc("amount", "the amount to build", Int::bounded(1, max)),
                        AfterSpace::new(OneOf::new(target_opts)),
                    ),
                    |(amount, target): (i32, BuildTarget)| Command::Build { amount, target },
                )))
            } else {
                None
            };

        // BuildTargetShipParser: `<amount> ship`.
        let ship_variant: Option<Box<dyn Parser<T = Command>>> = if self.can_build_ship(player) {
            let wood = b.goods.get(&Good::Wood).copied().unwrap_or(0);
            let cloth = b.goods.get(&Good::Cloth).copied().unwrap_or(0);
            let max = wood.max(cloth);
            Some(Box::new(Map::new(
                Chain2::new(
                    Doc::name_desc(
                        "amount",
                        "the amount of ships to build",
                        Int::bounded(1, max),
                    ),
                    AfterSpace::new(Token::new("ship")),
                ),
                |(amount, _): (i32, String)| Command::Build {
                    amount,
                    target: BuildTarget::Ship,
                },
            )))
        } else {
            None
        };

        let mut opts: Vec<Box<dyn Parser<T = Command>>> = vec![];
        if let Some(p) = worker_variant {
            opts.push(p);
        }
        if let Some(p) = ship_variant {
            opts.push(p);
        }

        Map::new(
            Chain2::new(
                Doc::name_desc(
                    "build",
                    "build a city, monument or ship",
                    Token::new("build"),
                ),
                AfterSpace::new(OneOf::new(opts)),
            ),
            |(_, cmd): (String, Command)| cmd,
        )
    }

    /// Port of `TradeParser`.
    fn trade_parser(&self, player: usize) -> impl Parser<T = Command> {
        let max = self.boards[player]
            .goods
            .get(&Good::Stone)
            .copied()
            .unwrap_or(0);
        Map::new(
            Chain2::new(
                Doc::name_desc(
                    "trade",
                    "trade stone for 3 workers each",
                    Token::new("trade"),
                ),
                AfterSpace::new(Doc::name_desc(
                    "amount",
                    "the amount of stone to trade",
                    Int::bounded(1, max),
                )),
            ),
            |(_, amount): (String, i32)| Command::Trade { amount },
        )
    }

    /// Port of `BuyParser`/`BuyDevelopmentParser`/`BuyGoodParser`.
    fn buy_parser(&self, player: usize) -> impl Parser<T = Command> {
        let available = self.available_developments(player);
        let choices: Vec<Choice<DevelopmentId>> = available
            .iter()
            .map(|&d| Choice {
                name: d.value().name,
                value: d,
            })
            .collect();
        let development_parser = Doc::name_desc(
            "development",
            "the development to buy",
            Enum::partial(choices),
        );

        let all_variant: Box<dyn Parser<T = BuyGoods>> =
            Box::new(Map::new(Token::new("all"), |_| BuyGoods {
                all_goods: true,
                goods: vec![],
            }));
        let goods_variant: Box<dyn Parser<T = BuyGoods>> = Box::new(Map::new(
            Many::bounded_spaced(good_parser(), 1, GOODS.len()),
            |goods: Vec<Good>| BuyGoods {
                all_goods: false,
                goods,
            },
        ));
        let goods_parser = OneOf::new(vec![all_variant, goods_variant]);

        Map::new(
            Chain3::new(
                Doc::name_desc("buy", "buy a development", Token::new("buy")),
                AfterSpace::new(development_parser),
                Opt::new(AfterSpace::new(goods_parser)),
            ),
            |(_, development, goods): (String, Choice<DevelopmentId>, Option<BuyGoods>)| {
                Command::Buy {
                    development: development.value,
                    goods: goods.unwrap_or_default(),
                }
            },
        )
    }

    /// Port of `TakeParser`.
    fn take_parser(&self) -> impl Parser<T = Command> {
        let max = self
            .kept_dice
            .iter()
            .filter(|&&d| d == crate::dice::Die::FoodOrWorkers)
            .count();
        let choices = vec![
            Choice {
                name: "food",
                value: TakeAction::Food,
            },
            Choice {
                name: "workers",
                value: TakeAction::Workers,
            },
        ];
        Map::new(
            Chain2::new(
                Doc::name_desc("take", "take food or workers from dice", Token::new("take")),
                AfterSpace::new(Many::bounded_spaced(Enum::partial(choices), 1, max)),
            ),
            |(_, actions): (String, Vec<Choice<TakeAction>>)| Command::Take {
                actions: actions.into_iter().map(|c| c.value).collect(),
            },
        )
    }

    /// Port of `DiscardParser`.
    fn discard_parser(&self, player: usize) -> impl Parser<T = Command> {
        let max = self.boards[player].goods_over_limit();
        Map::new(
            Chain3::new(
                Doc::name_desc(
                    "discard",
                    "discard goods down to the limit",
                    Token::new("discard"),
                ),
                AfterSpace::new(Doc::name_desc(
                    "amount",
                    "amount of goods to discard",
                    Int::bounded(1, max),
                )),
                AfterSpace::new(Doc::name_desc(
                    "good",
                    "type of good to discard",
                    good_parser(),
                )),
            ),
            |(_, amount, good): (String, i32, Good)| Command::Discard { amount, good },
        )
    }

    /// Port of `InvadeParser`.
    fn invade_parser(&self, player: usize) -> impl Parser<T = Command> {
        let max = self.boards[player]
            .goods
            .get(&Good::Spearhead)
            .copied()
            .unwrap_or(0);
        Map::new(
            Chain2::new(
                Doc::name_desc(
                    "invade",
                    "invade other players by spending spearheads for -2 points each",
                    Token::new("invade"),
                ),
                AfterSpace::new(Doc::name_desc(
                    "amount",
                    "amount of spearheads to spend",
                    Int::bounded(1, max),
                )),
            ),
            |(_, amount): (String, i32)| Command::Invade { amount },
        )
    }

    /// Port of `RollParser`.
    fn roll_parser(&self) -> impl Parser<T = Command> {
        let max_i = self.rolled_dice.len() as i32;
        let max_u = if self.phase == crate::Phase::ExtraRoll {
            1
        } else {
            max_i as usize
        };
        Map::new(
            Chain2::new(
                Doc::name_desc("roll", "roll dice", Token::new("roll")),
                AfterSpace::new(Doc::name_desc(
                    "dice",
                    "list of dice numbers to roll, separated by spaces",
                    Many::bounded_spaced(Int::bounded(1, max_i), 1, max_u),
                )),
            ),
            |(_, dice): (String, Vec<i32>)| Command::Roll { dice },
        )
    }

    /// Port of `SellParser`.
    fn sell_parser(&self, player: usize) -> impl Parser<T = Command> {
        let max = self.boards[player].food;
        Map::new(
            Chain2::new(
                Doc::name_desc("sell", "sell food for 6 coins each", Token::new("sell")),
                AfterSpace::new(Doc::name_desc(
                    "amount",
                    "amount of food to sell",
                    Int::bounded(1, max),
                )),
            ),
            |(_, amount): (String, i32)| Command::Sell { amount },
        )
    }
}

/// Port of `NextParser`.
fn next_parser() -> impl Parser<T = Command> {
    Map::new(
        Doc::name_desc(
            "next",
            "continue to the next phase of your turn",
            Token::new("next"),
        ),
        |_| Command::Next,
    )
}

/// Port of `PreserveParser`.
fn preserve_parser() -> impl Parser<T = Command> {
    Map::new(
        Doc::name_desc(
            "preserve",
            "use 1 pottery to double your food",
            Token::new("preserve"),
        ),
        |_| Command::Preserve,
    )
}

/// Port of `SwapParser`.
fn swap_parser() -> impl Parser<T = Command> {
    Map::new(
        Chain4::new(
            Doc::name_desc(
                "swap",
                "swap one type of goods for a different type using shipping",
                Token::new("swap"),
            ),
            AfterSpace::new(Doc::name_desc(
                "amount",
                "amount of goods to swap",
                Int::bounded(1, i32::MAX),
            )),
            AfterSpace::new(Doc::name_desc(
                "from",
                "type of good to swap away",
                good_parser(),
            )),
            AfterSpace::new(Doc::name_desc("to", "type of good to gain", good_parser())),
        ),
        |(_, amount, from, to): (String, i32, Good, Good)| Command::Swap { amount, from, to },
    )
}
